### DISCUSSION TITLE
How to make a more compact table?

### DISCUSSION THREAD

**Discussion by jkseppan** [quarto-dev/quarto-cli#2401] [Q&A]
In many cases, I find that the Bootstrap default `table` class has too much whitespace. The Bootstrap class `table-sm` would suit better, but is there any way to use that class for Markdown tables?

I found [a Pandoc discussion](https://stackoverflow.com/questions/41877612/pandoc-add-class-to-table-in-markdown) where it is mentioned that Pandoc doesn't currently have a way to specify table classes in Markdown, and that the easiest route would be to post-process the HTML. Indeed the following solution (run after quarto) is somewhat hacky but works, but I would prefer something that lets me control the class in the Markdown source.

    perl -pi -e 's/<table class="table">/<table class="table-sm">/g' _book/*.html

**Comment by mcanouil** [comment]
Customise the CSS for your own needs.

**Reply by jkseppan** [reply]
Sure, but how do I mark which tables should get the different CSS? I don't generally want to override every table (which my Perl solution does), I want to say that some specific tables should be more compact.

**Reply by mcanouil** [reply]
It will depends on how you generated the table.
It could be with a div cell, modifying the output of a code cell using `class-output` and/or `attr-output`, using (if R) the functions/options of `knitr`/`kableExtra`, `gt`, same for Python libraries that generate tables.

For a more specific answer, you are going to need to provide a small reproducible example of your use case.

**Reply by jkseppan** [reply]
I generated my table by writing a Markdown pipe table in a text editor. Here is a small example:

```
|   |   |   |   |   |    |    |    |    |    |   |   |   |   |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
|   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
|   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
|   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
|   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
|   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
|   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
```

I tried making a Lua filter to change the table class inside a specially marked-up div. This works when running just Pandoc:

```lua
-- table-class.lua
function Div(div)
  klass = div.attributes["table-class"]
  if klass ~= nil then
    return div:walk {
      Table = function(tbl)
        tbl.classes = { klass }
        return tbl
      end
    }
  end
end
```

With this, I would modify the source as follows:

```
:::{table-class=table-sm}
|   |   |   |   |   |    |    |    |    |    |   |   |   |   |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
|   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
|   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
|   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
|   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
|   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
|   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
:::
```

In raw Pandoc output, this generates a table with class `table-sm`. But if I list this filter in a Quarto document:

```
---
title: just testing
filters:
- table-class.lua
---

:::{table-class=table-sm}
|   |   |   |   |   |    |    |    |    |    |   |   |   |   |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
|   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
|   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
|   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
|   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
|   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
|   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
:::
```

...the HTML output has **both** `table` and `table-sm` classes. Apparently Quarto adds the `table` class after my filter is run.

**Comment by mcanouil** [comment]
- `table.lua`:
   ```lua
   local function has_value (tab, val)
       for index, value in ipairs(tab) do
           if value == val then
               return true
           end
       end
 
       return false
   end
 
   function Div(div)
     if has_value(div.classes, "table-sm") then
       return div:walk {
         Table = function(tbl)
           tbl.attributes['class'] = "table-sm"
           return tbl
         end
       }
     end
   end
   ```
- `index.qmd`
   ```
   ---
   title: "Untitled"
   format: html
   filters:
     - table.lua
   ---
    
   :::{.table-sm}
    
   |   |   |   |   |   |    |    |    |    |    |   |   |   |   |
   |:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
   |   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
   |   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
   |   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
   |   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
   |   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
   |   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
   |   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
    
   :::
    
   |   |   |   |   |   |    |    |    |    |    |   |   |   |   |
   |:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
   |   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
   |   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
   |   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
   |   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
   |   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
   |   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
   |   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
    
   ```
   
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/189744499-0507fcae-99aa-4336-a6e9-b65937636a28.png">

**Reply by jkseppan** [reply]
That does the same as my code above: the HTML table ends up with **both** `table` and `table-sm`. This reduces the vertical whitespace a little, but I'm looking for a way to **replace** `table` with `table-sm`. For reference, this is how the HTML rendering changes in that case:

<img width="824" alt="image" src="https://user-images.githubusercontent.com/92814/189946893-af6f43bd-123c-4211-ac9f-34b20d1bf33a.png">

**Reply by mcanouil** [reply]
By rewriting a bit the filter (see below), it appears the "table" class is added by Quarto after.
https://github.com/quarto-dev/quarto-cli/blob/main/src/format/html/format-html-bootstrap.ts#L299-L344

I don't see how (without using JavaScript) to remove the "table" class on load to change this behaviour.

```lua
local function has_value (tab, val)
    for index, value in ipairs(tab) do
        if value == val then
            return true
        end
    end

    return false
end

function Div(div)
  if has_value(div.classes, "table-sm") then
    return div:walk {
      Table = function(tbl)
        tbl.classes = pandoc.List()
        tbl.classes:insert("table-sm")
        quarto.utils.dump(tbl.classes)
        return tbl
      end
    }
  end
end
```

**Reply by jkseppan** [reply]
This is where Quarto adds the `table` class: https://github.com/quarto-dev/quarto-cli/blob/5ad49f1/src/format/html/format-html-bootstrap.ts#L299-L319

Because this happens in a postprocessing step, there is no way to prevent it with a Pandoc filter. I wonder if Quarto would accept a PR to prevent the change if the table already has some class attribute, or with a configuration option.

**Reply by jkseppan** [reply]
I suppose it would be possible to remove the classes in a post-render script similar to the Perl script I posted, but that would feel like a pretty convoluted way to do it.

**Reply by mcanouil** [reply]
I figured what post-processing script was involved (it was in my reply 😅)

Alternatives:

1. You can define `.table.table-sm` CSS rule strictly as `.table-sm`.
2. You can use a post-render script in Quarto, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>

**Reply by cderv** [reply]
> I wonder if Quarto would accept a PR to prevent the change if the table already has some class attribute, or with a configuration option.

I think this is reasonable. You could open an issue about this and with your PR suggestion.

**Reply by a0th** [reply]
> * `table.lua`:
>       ```lua
>       local function has_value (tab, val)
>           for index, value in ipairs(tab) do
>               if value == val then
>                   return true
>               end
>           end
>       
>           return false
>       end
>       
>       function Div(div)
>         if has_value(div.classes, "table-sm") then
>           return div:walk {
>             Table = function(tbl)
>               tbl.attributes['class'] = "table-sm"
>               return tbl
>             end
>           }
>         end
>       end
>       ```
> 
>     * `index.qmd`
>       ```
>       ---
>       title: "Untitled"
>       format: html
>       filters:
>         - table.lua
>       ---
>        
>       :::{.table-sm}
>        
>       |   |   |   |   |   |    |    |    |    |    |   |   |   |   |
>       |:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
>       |   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
>       |   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
>       |   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
>       |   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
>       |   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
>       |   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
>       |   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
>        
>       :::
>        
>       |   |   |   |   |   |    |    |    |    |    |   |   |   |   |
>       |:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
>       |   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
>       |   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
>       |   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
>       |   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
>       |   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
>       |   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
>       |   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
>        
>       ```
> 
> 
> <img alt="image" width="1624" src="https://user-images.githubusercontent.com/8896044/189744499-0507fcae-99aa-4336-a6e9-b65937636a28.png">

will this work for PDF?

**Reply by mcanouil** [reply]
No because it has nothing to do with LaTeX.

**Reply by a0th** [reply]
How could I make a PDF table more compact?

**Reply by mcanouil** [reply]
I am not sure to follow.
LaTeX tables are usually pretty much compact by default, there is no cell padding/margin contrary to bootstrap CSS / HTML.
Anyway, for LaTeX, to change the table layout you need to use LaTeX code: packages, options, etc.

**Comment by jkseppan** [comment]
This seems to work:

```lua
-- table-class.lua
function Div(div)
  klass = div.attributes["table-class"]
  if klass ~= nil then
    return div:walk {
      Table = function(tbl)
         tbl.classes = { klass }
         tbl.attributes.style = ""
        return tbl
      end,
      Tr = function(tr)
        tr.classes.remove("odd")
        tr.classes.remove("even")
        return tr
      end
    }
  end
end
```

```md
---
title: just testing
filters:
- table-class.lua
---

|   |   |   |   |   |    |    |    |    |    |   |   |   |   |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
|   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
|   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
|   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
|   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
|   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
|   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |

:::{table-class="table-sm"}
|   |   |   |   |   |    |    |    |    |    |   |   |   |   |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|   |   |   |   |   |    |    | 1  |    |    |   |   |   |   |
|   |   |   |   |   |    | 1  |    | 1  |    |   |   |   |   |
|   |   |   |   |   | 1  |    | 2  |    | 1  |   |   |   |   |
|   |   |   |   | 1 |    | 3  |    | 3  |    | 1 |   |   |   |
|   |   |   | 1 |   | 4  |    | 6  |    | 4  |   | 1 |   |   |
|   |   | 1 |   | 5 |    | 10 |    | 10 |    | 5 |   | 1 |   |
|   | 1 |   | 6 |   | 15 |    | 20 |    | 15 |   | 6 |   | 1 |
:::
```

The trick is to remove the `odd` and `even` classes from `tr` elements inside tables that have a class added. That way the table escapes modification by this line:

https://github.com/quarto-dev/quarto-cli/blob/5ad49f1/src/format/html/format-html-bootstrap.ts#L308

However, then the table has a style attribute added by Pandoc that defeats the styling. Ordinarily it is removed by that postprocessor but now I need to do it myself.

**Reply by jkseppan** [reply]
No, that doesn't work - I must have been running a modified version of Quarto when trying that out, since I was also trying to prove to myself that it is that code that is adding the class. The `odd` class is added by [Pandoc's HTML writer](https://hackage.haskell.org/package/pandoc-2.19.2/docs/src/Text.Pandoc.Writers.HTML.html#tableRowToHtml), and I don't think there is any option to override it from Quarto.
